spectrum-offchain-lm/src/executor.rs

```diff
@@ -128,7 +128,23 @@ where
         };
         if let Some(ord) = next_ord {
             trace!(target: "offchain_lm", "Order acquired [{:?}]", ord.get_self_ref());
-            let entity_id = ord.get_entity_ref();
+
+            // Note: if we're dealing with a Redeem order, then the PoolId must be obtained from
+            // its associated staking bundle. We cannot call `ord.get_entity_ref()` here; app will
+            // panic!
+            let entity_id = if let Order::Redeem(ref r) = ord {
+                let bundle_repo = self.bundle_repo.lock().await;
+                let bundle_id = BundleId::from(r.1.bundle_key.token_id);
+                trace!(target: "offchain_lm", "Requesting bundle with id [{:?}]", bundle_id);
+                if let Some(b) = bundle_repo.get_last_confirmed(bundle_id).await {
+                    b.0 .1.pool_id
+                } else {
+                    trace!(target: "offchain_lm", "Redeem order #{:?} is out of date", r.1.get_self_ref());
+                    return Err(());
+                }
+            } else {
+                ord.get_entity_ref()
+            };
```

Comment:
That's not a way to go. `entity_id` should be acquired before OnChainOrder is instantiated.

Reply:
Fixed. This is how I did it:
 - `Redeem` now contains `pool_id` again.
 - Create `RedeemProto` struct which doesn't contain pool id.
 - Create `OrderProto` which is the same as `Order` except it wraps `RedeemProto` instead of `Redeem`.
 - `OrderUpdatesHandler` emits `OrderProto` instead of `Order`
 - `backlog_stream` takes the incoming `OrderProto` instances and converts `RedeemProto` into `Redeem`.
   - `backlog_stream` needs access to `BundleRepo` instance, and so it was modified and also moved from `spectrum-offchain` to `spectrum-offchain-lm`.
 - Split `OnChainOrder` into `OnChainOrderId` and `OnChainEntityId` traits. Need this because `RedeemProto` can't impl `OnChainEntityId`.

Reply:
Nice! One suggestion: you can use `Has<OrderId>` for `RedeemProto` to avoid introducing new traits, then for regular `Redeem` we can leave `OnChainOrder`.

Reply:
Sure, it's done.